2023/07/wtchangdm/day7.rs

```diff
@@ -0,0 +1,213 @@
+        let mut map = HashMap::new();
+        for label in labels {
+            *map.entry(label).or_insert(0) += 1;
+        }
+
+        Self::inject_joker(&mut map);
+
+        let mut values: Vec<_> = map.values().collect();
+        values.sort();
+
+        match values.len() {
+            1 => Self::FiveOfAKind,
+            4 => Self::OnePair,
+            5 => Self::HighCard,
+            2 => match values[0..2] {
+                [1, 4] => Self::FourOfAKind,
+                [2, 3] => Self::FullHouse,
+                _ => panic!("how is this possible? 😑"),
+            },
+            3 => match values[0..3] {
+                [1, 1, 3] => Self::ThreeOfAKind,
+                [1, 2, 2] => Self::TwoPair,
+                _ => panic!("how is this possible? 😑"),
+            },
+            _ => panic!("how is this possible? 😑"),
+        }
+    }
+
+    fn rank(&self) -> u8 {
+        match self {
+            HandType::FiveOfAKind => 7,
+            HandType::FourOfAKind => 6,
+            HandType::FullHouse => 5,
+            HandType::ThreeOfAKind => 4,
+            HandType::TwoPair => 3,
+            HandType::OnePair => 2,
+            HandType::HighCard => 1,
+        }
+    }
+}
+
+struct Hand {
+    hand_type: HandType,
+    labels: Vec<u8>,
+    bid: u32,
+    use_joker: bool,
+}
+
+impl Hand {
+    fn parse_label(ch: &char, use_joker: bool) -> Label {
+        match ch {
+            'A' => 14,
+            'K' => 13,
+            'Q' => 12,
+            _ if *ch == 'J' && !use_joker => 11,
+            _ if *ch == 'J' && use_joker => 1,
+            'T' => 10,
+            '2'..='9' => *ch as u8 - b'0',
+            _ => panic!("how is this possible? 😑"),
+        }
+    }
+
+    fn from(input: &str, use_joker: bool) -> Self {
+        let (labels, bid) = input.split_once(' ').unwrap();
+        let labels: Vec<u8> = labels
+            .chars()
+            .map(|ch| Self::parse_label(&ch, use_joker))
+            .collect();
+        let hand_type = HandType::from(&labels);
+        let bid = bid.parse::<u32>().unwrap();
+
+        Self {
+            labels,
+            hand_type,
+            bid,
+            use_joker,
+        }
+    }
+}
+
+impl PartialOrd for Hand {
+    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
+        Some(self.cmp(other))
+    }
+}
+
+impl PartialEq for Hand {
+    fn eq(&self, other: &Self) -> bool {
+        self.cmp(other) == Ordering::Equal
+    }
+}
+
+impl Ord for Hand {
+    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
+        let result = self.hand_type.rank().cmp(&other.hand_type.rank());
+
+        if result != Ordering::Equal {
+            return result;
+        }
+
+        for (self_label, other_label) in self.labels.iter().zip(other.labels.iter()) {
+            let label_cmp = self_label.cmp(other_label);
+            if label_cmp != Ordering::Equal {
+                return label_cmp;
+            }
+        }
```

Comment:
here is [a default `Ord` impl for `Vec`](https://doc.rust-lang.org/1.74.1/std/vec/struct.Vec.html#impl-Ord-for-Vec%3CT,+A%3E), so should be sufficient to this case?

Something like

```
    fn cmp(…) -> Ordering {
        (self.hand_type.rank(), self.labels).cmp(&(other.hand_type.rank(), other.labels))
    }
```

Reply:
This is actually cool. Never thought it can work this way (also not sure where to find good examples).